Source/DafnyCore/Resolver/PreType/PreTypeResolve.State.cs

```diff
@@ -0,0 +1,589 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Diagnostics.Contracts;
+
+namespace Microsoft.Dafny {
+  public class PreTypeInferenceState {
```

Comment:
This class does more than hold state. Consider renaming to `PreTypeInferencer` ?

Could you add a piece of prose in a summary comment above it that explains some of its capabilities?

Reply:
I renamed it (and the file) `PreTypeConstraints` and added a summary description.